src/source/src/table.rs

```diff
@@ -67,75 +69,134 @@ impl TableDmlHandle {
+        if guard.changes_txs.is_empty() {
+            return Err(RwError::from(anyhow!(
+                "no available table reader in streaming source executors"
+            )));
         }
+        let len = guard.changes_txs.len();
+        let tx = guard
+            .changes_txs
+            .get((txn_id % len as u64) as usize)
+            .context("no available table reader in streaming source executors")?
+            .tx
+            .clone();
+
+        Ok(WriteHandle::new(txn_id, tx))
     }
 
     /// Get the reference of all columns in this table.
     pub(super) fn column_descs(&self) -> &[ColumnDesc] {
         self.column_descs.as_ref()
     }
+
+    pub fn check_chunk_schema(&self, chunk: &StreamChunk) {
+        risingwave_common::util::schema_check::schema_check(
+            self.column_descs
+                .iter()
+                .filter_map(|c| (!c.is_generated()).then_some(&c.data_type)),
+            chunk.columns(),
+        )
+        .expect("table source write txn_msg schema check failed");
+    }
 }
 
-#[easy_ext::ext(TableDmlHandleTestExt)]
-impl TableDmlHandle {
-    /// Write a chunk and assert that the chunk channel is not blocking.
-    fn write_chunk_ready(&self, chunk: StreamChunk) -> Result<oneshot::Receiver<usize>> {
-        self.write_chunk(chunk).now_or_never().unwrap()
+#[derive(Debug, PartialEq)]
+enum TxnState {
+    Init,
+    Begin,
+    Committed,
+    Rollback,
+}
+
+/// [`WriteHandle`] writes its data into a table in a transactional way.
+///
+/// First, it needs to call `begin()` and then write chunks by calling `write_chunk()`.
+///
+/// Finally call `end()` to commit the transaction or `rollback()` to rollback the transaction.
+///
+/// If the [`WriteHandle`] is dropped with a `Begin` transaction state, it will automatically
+/// rollback the transaction.
+pub struct WriteHandle {
+    txn_id: TxnId,
+    tx: mpsc::UnboundedSender<(TxnMsg, oneshot::Sender<usize>)>,
+    // Indicate whether `TxnMsg::End` or `TxnMsg::Rollback` have been sent to the write channel.
+    txn_state: TxnState,
+}
+
+impl Drop for WriteHandle {
+    fn drop(&mut self) {
+        if self.txn_state == TxnState::Begin {
+            let _ = self.rollback();
+        }
+    }
+}
+
+impl WriteHandle {
+    pub fn new(txn_id: TxnId, tx: mpsc::UnboundedSender<(TxnMsg, oneshot::Sender<usize>)>) -> Self {
+        Self {
+            txn_id,
+            tx,
+            txn_state: TxnState::Init,
+        }
+    }
+
+    pub fn begin(&mut self) -> Result<oneshot::Receiver<usize>> {
+        assert_eq!(self.txn_state, TxnState::Init);
+        self.txn_state = TxnState::Begin;
+        self.write_txn_msg(TxnMsg::Begin(self.txn_id))
+    }
+
+    pub fn write_chunk(&self, chunk: StreamChunk) -> Result<oneshot::Receiver<usize>> {
+        assert_eq!(self.txn_state, TxnState::Begin);
+        self.write_txn_msg(TxnMsg::Data(self.txn_id, chunk))
+    }
+
+    pub fn end(&mut self) -> Result<oneshot::Receiver<usize>> {
```

Comment:
Since the batch DML executors always send data (or control messages) to the same channel, I guess we can now only return the `oneshot` channel for the `End` message, which notifying us when the whole transaction is proceeded.

Reply:
Yes, only a single `oneshot` channel for a transaction is enough, even though we have `oneshot` channel for every messages.